### DISCUSSION TITLE
Integration of class-validator in the project DTOs

### DISCUSSION THREAD

**Discussion by coby111** [CADELmx/GPD_BACKEND#9] [Ideas]
# Idea to Integrate `class-validator` into the Project DTOs
## Description
I would like to suggest integrating the `class-validator` library into the DTOs of the project developed with NestJS. This library offers declarative validations through decorators in classes, which can simplify and improve the code.

## Benefits
- **Declarative Validations**: Allows defining validation rules directly on class properties, making the code cleaner and easier to understand.
- **Improved Maintainability**: Facilitates the management and modification of validation rules.
- **Security**: Ensures that data is validated before processing, reducing the risk of invalid or malicious data reaching the database.
- **Data Integrity**: Ensures that data meets necessary requirements before being inserted into the database, preventing integrity issues in the system.
- **Data Verification**: Allows verifying the validity of data before it is stored or used, ensuring the system handles only valid and correct data.
- **Seamless Integration with NestJS**: `class-validator` integrates smoothly with NestJS and its validation mechanisms.

## Example of Use
Here's an example of how we could use `class-validator` in a DTO:
```
import { IsString, IsInt, IsOptional } from 'class-validator';

export class CreateUserDto {
 @IsString()
 name: string;

 @IsInt()
 age: number;

 @IsOptional()
   @IsString()
  address?: string;
}
```

#### Install `class-validator` and `class-transformer`:
```
npm install class-validator class-transformer
```

### Update the DTOs:
Incorporate the necessary validations in the DTOs using the `class-validator` decorators.

### Update Controllers and Services:
Ensure that drivers use DTOs with validations. NestJS will automatically validate inputs based on the rules defined in the DTOs.

### Perform Tests:
Verify that validations work correctly and do not affect existing functionality.

## Comments
I am interested in receiving opinions and suggestions on this idea. What do you think about incorporating `class-validator` in our project?

## Resources
- [Documentation `class-validator`](https://github.com/typestack/class-validator)
- [Documentation `class-transformer`](https://github.com/typestack/class-transformer) 
- [Validation Guide in NestJS](https://docs.nestjs.com/techniques/validation)

**Comment by DiegoSHS** [comment]
I see that is a good idea of using validations following the [Validation Guide in NestJS](https://docs.nestjs.com/techniques/validation)
Maybe using auto-validation like the guide said is efficient, but there's some points that we have to consider
## Custom messages in responses
The class validator auto generates a response with messages in english, but the bussiness rules that we are implementing explicity say that we have to send responses with messages in spanish
there are some differences in the structure of the response that we are implementing and the auto generated responses that NestJS provide
- **Custom response structure in the project**
[GitHub - herramientas](https://trello.com/c/V0NQigPL/10-github )
```js
{
  “message“: ”ejemplo de mensaje” || null,
  “error“: null || {
    "message":"motivo del error"
  },
  “data“: […datos] || [ ] || null
}
```
- **Response generated by NestJS using class validator**
[Validation | NesJS - A progressive Node.js framework](https://docs.nestjs.com/techniques/validation#auto-validation)
```js
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": ["email must be an email"]
}
```
### How we can handle this
Notice that we can modify our custom structure of the responses to be inline with the structure of the auto generated responses that NestJS creates, it can be like this:
```js
{
  “message“:"Esta acción no se puede completar",
  “error“: "Error inesperado",
  “data“: null
}
```
_note: this is an example of a response on error_
I donk know if we can create custom messages using class validator in NestJS, let me know if it's possible
otherwise we can disable the detailed errors messages

**Reply by coby111** [reply]
# How to address handling custom messages in Spanish in NestJS responses

## Steps to follow

### Step 1: Set up custom messages

Configure custom messages in DTOs using `class-validator`.

```typescript
import { IsString, IsInt, IsOptional } from 'class-validator';

export class CreateTemplateDto {
  @IsString({ message: 'El nombre debe ser una cadena de caracteres' })
  name: string;

  @IsInt({ message: 'La edad debe ser un número entero' })
  age: number;

  @IsOptional()
  @IsString({ message: 'La dirección debe ser una cadena de caracteres' })
  address?: string;
}
```

### Step 2: Create a custom exception filter

Create a custom exception filter to capture and format error responses according to the desired structure.
The filter file can be located in `src/common/filters/custom-exception.filter.ts` or `src/models/filters/custom-exception.filter.ts`

```typescript
import { ExceptionFilter, Catch, ArgumentsHost, BadRequestException, HttpException } from '@nestjs/common';
import { Response } from 'express';

@Catch(HttpException)
export class CustomExceptionFilter implements ExceptionFilter {
  catch(exception: HttpException, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<Response>();
    const status = exception.getStatus();
    const exceptionResponse = exception.getResponse() as any;

    let message = 'Error inesperado';
    let errorMessage = 'Esta acción no se puede completar';

    if (exception instanceof BadRequestException && exceptionResponse.message) {
      errorMessage = exceptionResponse.message.join(', ');
      message = 'Solicitud incorrecta';
    } else if (typeof exceptionResponse === 'object' && exceptionResponse.message) {
      errorMessage = exceptionResponse.message;
    }

    response.status(status).json({
      message: errorMessage,
      error: {
        message,
      },
      data: null,
    });
  }
}
```
### Or also only modify the http responses, without modifying the DTOs:
```typescript
import { ExceptionFilter, Catch, ArgumentsHost, HttpException, HttpStatus } from '@nestjs/common';
import { Request, Response } from 'express';

@Catch(HttpException)
export class HttpExceptionFilter implements ExceptionFilter {
  catch(exception: HttpException, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<Response>();
    const request = ctx.getRequest<Request>();
    const status = exception.getStatus();

    // Map status codes to Spanish messages
    const errorMessages = {
      [HttpStatus.BAD_REQUEST]: 'Solicitud incorrecta',
      [HttpStatus.UNAUTHORIZED]: 'No autorizado',
      [HttpStatus.FORBIDDEN]: 'Prohibido',
      [HttpStatus.NOT_FOUND]: 'No encontrado',
      [HttpStatus.INTERNAL_SERVER_ERROR]: 'Error interno del servidor',
      // Add more status codes and messages as needed
    };

    const message = errorMessages[status] || 'Error desconocido';

    response.status(status).json({
      timestamp: new Date().toISOString(),
      path: request.url,
      error: {
        message,
      },
    });
  }
}
```

### Step 3: Apply the exception filter
Applies the exception filter to the application, either globally or on specific handlers.

- #### Apply globally

In the main file (main.ts):
```typescript
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app.module';
import { CustomExceptionFilter } from './common/filters/custom-exception.filter';

async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  app.useGlobalFilters(new CustomExceptionFilter());
  await app.listen(3000);
}
bootstrap();
```

- #### Apply to Specific Drivers
On the controller:

```typescript
import { Controller, Post, Get, Put, Delete, Param, Body, UseFilters } from '@nestjs/common';
import { TemplateService } from './template.service';
import { CreateTemplateDto, UpdateTemplateDto } from './dto/template.dto';
import { CustomExceptionFilter } from './common/filters/custom-exception.filter';

@Controller('templates')
@UseFilters(CustomExceptionFilter)
export class TemplateController {
  constructor(private readonly templateService: TemplateService) {}

  @Post()
  async create(@Body() createTemplateDto: CreateTemplateDto) {
    return this.templateService.create(createTemplateDto);
  }

  @Get()
  async findAll() {
    return this.templateService.findAll();
  }

  @Put(':id')
  async update(@Param('id') id: number, @Body() updateTemplateDto: UpdateTemplateDto) {
    return this.templateService.update(id, updateTemplateDto);
  }

  @Delete(':id')
  async remove(@Param('id') id: number) {
    return this.templateService.remove(id);
  }
}
```

With these steps, you can handle custom messages in Spanish and format error responses according to the desired structure in the project.

**Reply by DiegoSHS** [reply]
Nice, we can use both, implementing http filters to generate the custom message in spanish when there's an error, and the class validator will be provide the message of the detailed error.
## ValidationPipe + ExceptionFilter
We can use validationpipes to ensure all endpoints are protected from receiving incorrect data and the exception filters at the same time globally, i find this aproach more efficient to avoid applying to specific drivers
```js
async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  app.useGlobalFilters(new CustomExceptionFilter());
  app.useGlobalPipes(new ValidationPipe());
  await app.listen(3000);
}
bootstrap();
```
### The exception filter
```js
import { ExceptionFilter, Catch, ArgumentsHost, HttpException, HttpStatus } from '@nestjs/common';
import { Request, Response } from 'express';

@Catch(HttpException)
export class HttpExceptionFilter implements ExceptionFilter {
  catch(exception: HttpException, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<Response>();
    const status = exception.getStatus();

    // Map status codes to Spanish messages
    const errorMessages = {
      [HttpStatus.BAD_REQUEST]: 'Solicitud incorrecta',
      [HttpStatus.UNAUTHORIZED]: 'No autorizado',
      [HttpStatus.FORBIDDEN]: 'Prohibido',
      [HttpStatus.NOT_FOUND]: 'No encontrado',
      [HttpStatus.INTERNAL_SERVER_ERROR]: 'Error interno del servidor',
      // Add more status codes and messages as needed
    };

    const message = errorMessages[status] || 'Error desconocido';

    response.status(status).json({
      error: message,
    });
  }
}
```
## Conditional validation
We can use conditional validation for cases when we need to apply different validations in specific properties
```js
import { ValidateIf, IsNotEmpty } from 'class-validator';

export class Post {
  otherProperty: string;

  @ValidateIf(o => o.otherProperty === 'value')
  @IsNotEmpty({message: 'El nombre está vacío'})
  nombre: string;
}
```
## Expected result
With that configuration, the responses should look like this:
```js
{
  "error": "Solicitud incorrecta",
  "message": "El nombre está vacío"
}
```
_note: this is only an example of the expected body of the response on error, we have to perform tests before_

**Comment by elisa242003** [comment]
Based on the suggestions given and the answers to the questions asked, I consider that `class-validator` is a good option that would help us reduce time and effort in including validations for each module. I think we can still correct small changes so that what we develop is consistent with what we ask for.

**Comment by DiegoSHS** [comment]
Closing discussion, this will be added to our trello board and the implementation discussed for more improvements, i hope end well implemented
